editoast/src/views/timetable/train_schedule.rs

```diff
@@ -590,6 +609,309 @@ async fn occupancy_blocks(
+                number: missing.len(),
+            },
+        )
+        .await?;
+
+    let operational_point = OperationalPointModel::retrieve_real_or_fail(
+        db_pool.get().await?,
+        (infra_id, op_id.clone()),
+        || TrainScheduleError::OperationalPointNotFound {
+            op_id: op_id.clone(),
+        },
+    )
+    .await?;
+
+    // Retrieve simulations
+    let simulations_result = train_simulation_batch(
+        &mut db_pool.get().await?,
+        valkey,
+        core_client,
+        &train_schedules,
+        &infra,
+        electrical_profile_set_id,
+    )
+    .await?;
+
+    let operational_point_track_offsets = operational_point.schema.track_offset();
+
+    let path_items = train_schedules
+        .iter()
+        .flat_map(|ts| ts.path.iter().map(|p| &p.location))
+        .collect::<Vec<_>>();
+
+    let path_item_cache =
+        PathItemCache::load(&mut db_pool.get().await?, infra_id, &path_items).await?;
+
+    for ((simulation, pathfinding), train_schedule) in
+        simulations_result.iter().zip(train_schedules)
+    {
+        // Get the track ranges from the pathfinding
+        let (track_ranges, path_item_positions) = match pathfinding {
+            PathfindingResult::Success(pathfinding_result_success) => (
+                &pathfinding_result_success.track_section_ranges,
+                &pathfinding_result_success.path_item_positions,
+            ),
+            PathfindingResult::Failure(_) => {
+                tracing::info!(train_schedule.id, "pathfinding failed");
+                continue;
+            }
+        };
+
+        // Get the positions and the times from the simulation
+        let report_train = match simulation {
+            Response::Success { final_output, .. } => final_output.report_train.clone(),
+            _ => {
+                tracing::info!(train_schedule.id, "simulation failed");
+                continue;
+            }
+        };
+
+        let path_item_id = match_path_item_id_with_op_id(&path_item_cache, &train_schedule, &op_id);
+
+        if let Some((track_section, track_occupancy)) = compute_track_occupancy(
+            path_item_id,
+            path_item_positions,
+            track_ranges,
+            &operational_point_track_offsets,
+            &report_train,
+            &train_schedule,
+        ) {
+            track_occupancy_map
+                .entry(track_section)
+                .or_default()
+                .push(track_occupancy)
+        }
+    }
+    Ok(Json(track_occupancy_map))
+}
+
+/// Computes the occupancy of track sections for a given train based on its schedule and path.
+/// If a path item ID is provided, it uses schedule data to determine stop duration and location.
+/// If not, it infers the position and time based on track offsets and interpolation.
+fn compute_track_occupancy(
+    path_item_id: Option<NonBlankString>,
+    path_item_positions: &[u64],
+    track_ranges: &Vec<TrackRange>,
```

Comment:
There is almost never a good reason to take a vec by reference. 

```suggestion
    track_ranges: &[TrackRange],
```

or even better when random access isn't necessary

```suggestion
    track_ranges: impl Iterator<Item = TrackRange>,
```